### ISSUE TITLE
gulp-eslint wont process more than 16 files

### ISSUE THREAD

**Issue by raleksandar** [adametry/gulp-eslint#36]
I stumbled upon a very strange issue with gulp-eslint. I started working on a new project and as I kept adding new files eslint stopped reporting any issues at some point. 

So I tried to reduce test case and came up with following:

Gulpfile.js:
````js
var gulp = require('gulp'),
    debug = require('gulp-debug'),
    eslint = require('gulp-eslint');

gulp.task('lint', function() {
    gulp.src('./src/js/**/*.js')
        .pipe(debug())
        .pipe(eslint())
        .pipe(eslint.format());
});

gulp.task('default', ['lint']);
````
In ./src/js I have 16 identical files with following contents:
````js
'use strict';
{
````
If I run gulp I will get 16 "Unexpected end of input" errors (as I should):
````
λ gulp
[15:09:42] Using gulpfile E:\tmp\eslint-test\gulpfile.js
[15:09:42] Starting 'lint'...
[15:09:42] Finished 'lint' after 19 ms
[15:09:42] Starting 'default'...
[15:09:42] Finished 'default' after 7.6 μs
[15:09:42] gulp-debug: src\js\1.js
[15:09:42] gulp-debug: src\js\10.js
[15:09:42] gulp-debug: src\js\11.js
[15:09:42] gulp-debug: src\js\12.js
[15:09:42] gulp-debug: src\js\13.js
[15:09:42] gulp-debug: src\js\14.js
[15:09:42] gulp-debug: src\js\15.js
[15:09:42] gulp-debug: src\js\16.js
[15:09:42] gulp-debug: src\js\2.js
[15:09:42] gulp-debug: src\js\3.js
[15:09:42] gulp-debug: src\js\4.js
[15:09:42] gulp-debug: src\js\5.js
[15:09:43] gulp-debug: src\js\6.js
[15:09:43] gulp-debug: src\js\7.js
[15:09:43] gulp-debug: src\js\8.js
[15:09:43] gulp-debug: src\js\9.js
[15:09:43] gulp-debug: 16 items
[15:09:43]
E:\tmp\eslint-test\src\js\1.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\10.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\11.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\12.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\13.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\14.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\15.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\16.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\2.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\3.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\4.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\5.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\6.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\7.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\8.js
  4:1  error  Unexpected end of input

E:\tmp\eslint-test\src\js\9.js
  4:1  error  Unexpected end of input

✖ 16 problems (16 errors, 0 warnings)
````
But as soon as I add one more file:
````
cp src/js/1.js src/js/17.js
````

I no longer get any reports from eslint:
````
λ gulp
[15:10:11] Using gulpfile E:\tmp\eslint-test\gulpfile.js
[15:10:11] Starting 'lint'...
[15:10:11] Finished 'lint' after 23 ms
[15:10:11] Starting 'default'...
[15:10:11] Finished 'default' after 6.46 μs
[15:10:11] gulp-debug: src\js\1.js
[15:10:11] gulp-debug: src\js\10.js
[15:10:11] gulp-debug: src\js\11.js
[15:10:11] gulp-debug: src\js\12.js
[15:10:11] gulp-debug: src\js\13.js
[15:10:11] gulp-debug: src\js\14.js
[15:10:11] gulp-debug: src\js\15.js
[15:10:11] gulp-debug: src\js\16.js
[15:10:11] gulp-debug: src\js\17.js
[15:10:11] gulp-debug: src\js\2.js
[15:10:11] gulp-debug: src\js\3.js
[15:10:11] gulp-debug: src\js\4.js
[15:10:11] gulp-debug: src\js\5.js
[15:10:11] gulp-debug: src\js\6.js
[15:10:11] gulp-debug: src\js\7.js
[15:10:11] gulp-debug: src\js\8.js
[15:10:11] gulp-debug: src\js\9.js
[15:10:11] gulp-debug: 17 items
````
I have discovered that in latter case the flushFunction through2 parameter  in the gulpEslint.format function is not being called. I tried replacing through with Node's PassThrough:
````js
var stream = require('stream');

gulpEslint.format = function(formatter, writable) {
    var results = [];
    formatter = util.resolveFormatter(formatter);
    writable = util.resolveWritable(writable);

    return new  stream.PassThrough({objectMode: true})
        .on('data', function(file) {
            if (file.eslint) {
                results.push(file.eslint);
            }
        })
        .on('end', function() {
            if (results.length) {
                util.writeResults(results, formatter, writable);
            }
        });
};
````
And it worked, and keeps working as many new files I add.

Is this a valid fix? I can create a PR if it is..

**Comment by shinnn** [comment]
Interesting. Which version of Node are you using?

**Comment by raleksandar** [comment]
I'm little outdated (cannot upgrade due to some incompatibilities with company's internal projects):
````
λ node -v
v0.10.36
````

**Comment by raleksandar** [comment]
What I forgot to mention is that cli eslint works just fine (through2 was the culprit)

**Comment by shinnn** [comment]
@raleksandar Thanks for reporting. I'm investigating the cause on my machine... maybe I'll ask you to create a pull request later.

**Comment by Kingdutch** [comment]
:+1: Experienced the same issue on node 0.12.0

This is actually a major issue, I thought gulp-eslint didn't work at all as it gave no output for the project I added it to but eslint itself gave correct output.

**Comment by alexfedoseev** [comment]
Same here: there is the output with `eslint cli`, but there isn't with `gulp-eslint`.

**Comment by adametry** [comment]
Unfortunately, I'm having trouble reproducing this one (tried on: Win10/OSX10.10, Node@0.10.36/0.12.0, and through2@0.6.3/0.6.5). My test involved 16/17 identical files and a gulpfile according to raleksandar's description.

Regarding the issue, are you still experiencing this issue? If so, does passing a highWaterMark option for through2 change where that count fails. For example:

```javascript
gulpEslint.format = function(formatter, writable) {
	var results = [];
	formatter = util.resolveFormatter(formatter);
	writable = util.resolveWritable(writable);

	//  raise highWaterMark to 17...
	return through.obj({ highWaterMark:17 }, function(file, enc, cb) {
		if (file.eslint) {
			results.push(file.eslint);
		}
		cb(null, file);
	}, function(cb) {
		// Only format results if files has been lint'd
		if (results.length) {
			util.writeResults(results, formatter, writable);
		}
		// reset buffered results
		results = [];
		cb();
	});
};
```

If it still remains, I can replace through2 with a simple `stream.Transform`. A `stream.PassThrough` would take away the ability for `format` to be async.

**Comment by AlanFoster** [comment]
Also facing this issue; After 16 files all output is swallowed.

The solution offered by @raleksandar proved to work; Similarly the solution provided by @adametry to update `highWaterMark` to our current number of files worked too. This [issue](https://github.com/rvagg/through2/issues/32) may be relevant.

This is reproducible under 
- OSX 10.9.5
- Node v0.10.36

It might be nice to move forwards with replacing `through2` as a temporary solution, and to deal with any performance concerns in the future perhaps?

**Comment by juhasz** [comment]
Have the same problem.
Until a proper solution came up, I use a pretty simple workaround, which don't requires to change the plugins: just bind an empty function on 'data' events of the eslint functions in the task definition.

Like this:
```
gulp.task('eslint', function() {
  gulp.src(filesToLint)
   .pipe(eslint())
    // Binding a function on data event is a workaround to resolve https://github.com/adametry/gulp-eslint/issues/36
   .pipe(eslint.format().on('data', function() {}))
   .pipe(eslint.failAfterError().on('data', function() {}));
});
```

**Comment by adametry** [comment]
`through2` use has been replaced with a with `Transform` streams. This change has been published as v0.13.2.

Thanks all the assistance on this one! Let me know if any of you still experience this issue.

**Comment by raleksandar** [comment]
I've just upgraded to the 0.13.2 and run the same test from the first post in this issue and I still don't get output when more than 16 files is passed through the stream. 
Adding `highWaterMark: 17` to the options of the stream constructed in `util.transform()` allows for 17th file to be processed but not more. So the issue is still present.

I run this test on Node v0.12.0 under 64bit Windows 8.1

**Comment by adametry** [comment]
Oh man. I should have noticed this a long time ago...

Make sure to `return` the lint stream in the gulp task:
```javascript
var gulp = require('gulp'),
    debug = require('gulp-debug'),
    eslint = require('gulp-eslint');

gulp.task('lint', function() {
    // returning the stream will ensure that gulp keeps the process alive until it's done.
    // Otherwise, gulp considered this a sync task and moves on.
    return gulp.src('./src/js/**/*.js')
        .pipe(debug())
        .pipe(eslint())
        .pipe(eslint.format());
});

gulp.task('default', ['lint']);
```

I can reproduce the error easily when the `return` is removed.

if you look back at your output, you'll see that the gulp task finished before all the output was complete:
```bash
...
[15:10:11] Starting 'lint'...               ⬅︎ "lint" task dependency starting
                                            ⬅︎ This is where the gulp-debug and format output should appear
[15:10:11] Finished 'lint' after 23 ms      ⬅︎ "lint" dependency finishes before gulp.src sends a file
[15:10:11] Starting 'default'...            ⬅︎ W/ deps met, starting "default" (which is a no-op)
[15:10:11] Finished 'default' after 6.46 μs ⬅︎ Gulp is done. Anything after this is the output stream being flushed
[15:10:11] gulp-debug: src\js\1.js          ⬅︎ The "lint" task's stream is still running
...
```
The reason that gulp-debug is printed is that it's printing on "data"; however, `eslint.format` waits for the stream to complete so it can print a single lint message for all files (note: use eslint.formatEach to print independently on the "data" events).

When under the `highWatermark`, the stream is able to complete before gulp shuts down, albeit after the gulp task completes (async race). When the `highWatermark` is reached, stream buffering kicks in and delays the completion of the stream. In these cases, you'll occasionally see more or less of the lint task make it out before it ends (including missing some gulp-debug messages); however, the added delay makes it very unlikely that it'll reach the stream's end to print the lint message before exiting.

**tl;dr:** be sure to return the stream formed in a gulp task

(@juhasz's code has the same issue)

**Comment by raleksandar** [comment]
I can confirm that returning the stream works. But it seems that returning a Promise or using callback-style async tasks is not possible since `finish` event is never emitted on the stream.

For example, by reading the [Gulp docs](https://github.com/gulpjs/gulp/blob/master/docs/API.md#async-task-support) one could try the following:

```js
gulp.task('lint', function(cb) {
    gulp.src('./src/js/**/*.js')
        .pipe(eslint())
        .pipe(eslint.format())
        .on('finish', cb);
});

// or

gulp.task('lint', function() {
    return new Promise(function(fulfill, reject) {
        gulp.src('./src/js/**/*.js')
            .pipe(eslint())
            .pipe(eslint.format())
            .on('finish', fulfill)
            .on('error', reject);
    });
});
```

So maybe this should just be mentioned in the README especially since this is the first gulp plugin I've used that requires you to define an async task.

**Comment by adametry** [comment]
This has to do with how streams work. You can remove gulp-eslint or use any other plugin from your examples, and the outcome is the same (the process exits before either task is completed). In both examples, `"finish"` on the last stream in the pipeline will never emit because nothing reading it (either `pipe` to a writable stream or listen for `"data"` events). 

The reason that the lint results are not printed is that `format` buffers results and prints them on the `"prefinish"` event (which occurs just before the stream completes). This matches how messages are printed in eslint (multiple files, one summary). However, since there are cases where the report should be printed as the file passes through, `formatEach` is an alternative that prints as the data arrives.

If you use `formatEach` in the examples above, you'll see the lint results, but the tasks still will not end. Try adding `.on("data", function noop() {})` to that last stream (like you do with the `"finish"` event). This will let the last stream flow and emit the `"finish"` event.

I'll consider some ways to make this clear in the docs, but this is not unique to gulp-eslint and perhaps should be explained in the gulp docs.

**Comment by raleksandar** [comment]
Ok, now I understand why it works like that, thanks for a great explanation. And you're right, it's probably better if this was explained well in gulp docs but I do believe that a simple note in gulp-eslint readme about importance of returning a stream will help others to avoid making the same mistake as I did.

**Comment by gauntface** [comment]
Would it be possible to put a note on the readme regarding this issue? Specifically calling out the returning a promise use case?

We hit the same issue as @raleksandar and knowing about the noop data trick would have made it much easier to diagnose and fix.

I'm more than happy to raise a PR if you are open to the idea.

**Comment by WillsB3** [comment]
We also just hit this problem... Very strange and annoying.
